gson/src/main/java/com/google/gson/internal/reflect/ReflectionHelper.java

```diff
@@ -169,22 +177,34 @@ public <T> Constructor<T> getCanonicalRecordConstructor(Class<T> raw) {
       try {
         Object recordComponents = getRecordComponents.invoke(raw);
         int componentCount = Array.getLength(recordComponents);
-        Class[] recordFields = new Class[componentCount];
+        Class<?>[] recordComponentType = new Class<?>[componentCount];
         for (int i = 0; i < componentCount; i++) {
-          recordFields[i] =
-                  Class.class.cast(getType.invoke(Array.get(recordComponents, i)));
+          recordComponentType[i] = Class.class.cast(getType.invoke(Array.get(recordComponents, i)));
         }
-        return raw.getConstructor(recordFields);
-      } catch (IllegalAccessException e) {
-        throw createExceptionForUnexpectedIllegalAccess(e);
+        return raw.getDeclaredConstructor(recordComponentType);
       } catch (ReflectiveOperationException e) {
-        throw new RuntimeException("Failed to reflect into record components on"
-                + " class [" + raw + "], this is unexpected behaviour, as these methods should"
-                + " not throw when they are defined on Class",
-                e
-        );
+        throw createExceptionForRecordReflectionException(e);
       }
     }
+
+    @Override
+    public Method getAccessor(Class<?> raw, Field field) {
+      try {
+        Object recordComponents = getRecordComponents.invoke(raw);
```

Comment:
Would probably simplify the code below to cast this to `Object[]`.

Reply:
Done